src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -285,30 +285,29 @@ private[concurrent] final object Promise {
       }
 
     // IMPORTANT: Noop should never be passed in here, neither as left OR as right
-    @tailrec private[this] final def concatCallbacks(left: Callbacks[T], right: Callbacks[T]): Callbacks[T] = {
+    @tailrec private[this] final def concatCallbacks(left: Callbacks[T], right: Callbacks[T]): Callbacks[T] =
       if (left.isInstanceOf[Transformation[T,_]]) new ManyCallbacks[T](left, right)
       else /*if (left.isInstanceOf[ManyCallbacks[T]) */ { // This should only happen when linking
         val m = left.asInstanceOf[ManyCallbacks[T]]
         concatCallbacks(m.last, new ManyCallbacks(m.first, right))
       }
-    }
 
-    // IMPORTANT: Noop should not be passed in here
+    // IMPORTANT: Noop should not be passed in here, `callbacks` cannot be null
     @tailrec
     private[this] final def submitWithValue(callbacks: Callbacks[T],
                                             resolved: Try[T],
-                                            deferred: List[Callbacks[T]] = Nil): Unit = {
-      callbacks match {
-        case m: ManyCallbacks[T] =>
-          submitWithValue(m.first, resolved, m.last :: deferred)
-        case t: Transformation[T, _] =>
-          t.submitWithValue(resolved)
-          deferred match {
-            case head :: tail => submitWithValue(head, resolved, tail)
-            case Nil => //ignore
-          }
+                                            deferred: Callbacks[T] = null): Unit =
+      if(callbacks.isInstanceOf[ManyCallbacks[T]]) {
+        val m: ManyCallbacks[T] = callbacks.asInstanceOf[ManyCallbacks[T]]
```

Comment:
Is this faster than `case m: ManyCallbacks[T]`?

Reply:
pattern matching generates more bytecode which reduces inlineability at the callsite.

Reply:
That's unfortunate.  In my mini reproduction I'm seeing this bytecode diff:

```diff
diff --git Main.scala Main.scala
index 0a62047..13b5dd3 100644
--- Main.scala
+++ Main.scala
@@ -7,10 +7,10 @@ final class Transformation[-F, T] extends Callbacks[F]

 class Bar {
   def foo[T](callbacks: Callbacks[T]): Unit = {
-    if (callbacks.isInstanceOf[ManyCallbacks[T]]) {
-      val m: ManyCallbacks[T] = callbacks.asInstanceOf[ManyCallbacks[T]]
+    callbacks match {
+      case m: ManyCallbacks[T] =>
         foo(m.first)
-    } else {
+      case _ =>
         val t: Transformation[T, _] = callbacks.asInstanceOf[Transformation[T, _]]
         println(t.toString)
     }
diff --git t/Bar.class t/Bar.class
index 1d6f41b..345afc1 100644
--- t/Bar.class
+++ t/Bar.class
@@ -5,22 +5,30 @@ public class t.Bar {
     descriptor: (Lt/Callbacks;)V
     Code:
 ##: aload_1
+##: astore_3
+##: aload_3
 ##: instanceof    ###                 // class t/ManyCallbacks
 ##: ifeq          ##
-##: aload_1
+##: aload_3
 ##: checkcast     ###                 // class t/ManyCallbacks
-##: astore_2
-##: aload_2
+##: astore        ##
+##: aload         ##
 ##: invokevirtual ###                 // Method t/ManyCallbacks.first:()Lt/Callbacks;
 ##: invokevirtual ###                 // Method foo:(Lt/Callbacks;)V
+##: getstatic     ###                 // Field scala/runtime/BoxedUnit.UNIT:Lscala/runtime/BoxedUnit;
+##: astore_2
+##: goto          ##
 ##: goto          ##
 ##: aload_1
 ##: checkcast     ###                 // class t/Transformation
-##: astore_3
+##: astore        ##
 ##: getstatic     ###                 // Field scala/Predef$.MODULE$:Lscala/Predef$;
-##: aload_3
+##: aload         ##
 ##: invokevirtual ###                 // Method t/Transformation.toString:()Ljava/lang/String;
 ##: invokevirtual ###                 // Method scala/Predef$.println:(Ljava/lang/Object;)V
+##: getstatic     ###                 // Field scala/runtime/BoxedUnit.UNIT:Lscala/runtime/BoxedUnit;
+##: astore_2
+##: goto          ##
 ##: return

   public t.Bar();
```

Reply:
yeah，unfortunate，could the compiler smart enough to optimize this out?

Reply:
:man_shrugging:

Reply:
Not sure, but one would hope so. In the mean time, I'll have to manually write the optimal code. :/